### PR TITLE
Add pack contract implementation

### PR DESCRIPTION
**Pull request by adam-maj**

### CODE REVIEW

File: src/schema/tokens/pack.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  ERC721WrappableSchema,
+} from "./common/wrap";
+
+/**
+ * @internal
+ */
+const ERC20RewardSchema = ERC20WrappableSchema.omit({
+  quantity: true,
+}).extend({
+  quantityPerReward: PriceSchema,
```
Comment (joaquim-verges) [inline src/schema/tokens/pack.ts:16]: wonder if it's really worth it to rename this. would help with code sharing if we didnt

File: src/schema/tokens/pack.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  erc1155Rewards: z.array(ERC1155RewardSchema).default([]),
+});
+
+/**
+ * @internal
+ */
+export const PackRewardsOutputSchema = z.object({
+  erc20Rewards: z.array(ERC20RewardContentsSchema).default([]),
+  erc721Rewards: z.array(ERC721RewardContentsSchema).default([]),
+  erc1155Rewards: z.array(ERC1155RewardContentsSchema).default([]),
```
Comment (joaquim-verges) [inline src/schema/tokens/pack.ts:67]: same for this - `erc20Contents` would make sense in both the pack and MW context right?

File: test/pack.test.ts [hunk 3]
```diff
@@ -76,118 +82,103 @@ describe("Pack Contract", async () => {
...
-          amount: BigNumber.from(50),
+          quantityPerReward: 1,
+          totalRewards: 50,
         },
         {
+          contractAddress: bundleContract.getAddress(),
           tokenId: "2",
-          amount: BigNumber.from(50),
+          quantityPerReward: 1,
+          totalRewards: 50,
```
Comment (joaquim-verges) [inline test/pack.test.ts:102]: this does read kinda nice though

File: test/pack.test.ts [hunk 4]
```diff
@@ -76,118 +82,103 @@ describe("Pack Contract", async () => {
...
         },
       ],
       metadata: {
```
Comment (joaquim-verges) [inline test/pack.test.ts:105]: should we call this "packMetadata" to be clear its for the pack?

File: src/contracts/pack.ts [hunk 5]
```diff
@@ -121,448 +126,428 @@ export class Pack implements UpdateableNetwork {
...
+   * @param metadataWithRewards - the metadata and rewards to include in the pack
    *
    * @example
    * ```javascript
-   * // Data to create the pack
-   * const pack = {
-   *   // The address of the contract that holds the rewards you want to include
-   *   assetContract: "0x...",
-   *   // The metadata of the pack
+   * const packMetadata = {
```
Comment (joaquim-verges) [inline src/contracts/pack.ts:292]: i would call this `pack` and `metadata` -> `packMetadata`

**Review by joaquim-verges — COMMENTED** [review]
overall looks good, minor things - mainly trying to avoid some code duplication if we can. Though the API reads pretty nice right now

File: src/contracts/pack.ts [hunk 6]
```diff
@@ -49,17 +55,15 @@ import { ContractAnalytics } from "../core/classes/contract-analytics";
...
   static contractAbi = require("../../abis/Pack.json");
   /**
    * @internal
    */
-  static schema = PacksContractSchema;
-
-  private contractWrapper: ContractWrapper<PackContract>;
-  private storage: IStorage;
+  static schema = PackContractSchema;
+  // TODO: Change schema and deployment
```
[the remark this answers is not in the document]
Comment (adam-maj) [reply src/contracts/pack.ts:66]: Oops just forgot to remove the comment

File: src/contracts/pack.ts [hunk 7]
```diff
@@ -121,448 +126,428 @@ export class Pack implements UpdateableNetwork {
...
+      const hasAllowance = await hasERC20Allowance(
+        this.contractWrapper,
+        erc20.contractAddress,
+        totalQuantity,
+      );
+      if (!hasAllowance) {
+        throw new Error(
+          `ERC20 token with contract address "${
+            erc20.contractAddress
+          }" does not have enough allowance to transfer.\n\nYou can set allowance to the multiwrap contract to transfer these tokens by running:\n\nawait sdk.getToken("${
```
[the remark this answers is not in the document]
Comment (adam-maj) [reply src/contracts/pack.ts:473]: There are some key differences because of the pack interface we want vs the one the contract implements, we need to do some multiplication/division in all these in the pack file that's different for each function in pack + different from multiwrap - which is why I opted to keep these all separate as verbose as it is, b/c handling all the casing seems messier and wanted to keep things explicit.

TL;DR, because contract uses total tokens + number of rewards (division), but we use tokens per reward + number of rewards (multiplication) we have to do some conversions between these interfaces which makes things different from MW and each other.

File: src/schema/tokens/pack.ts (hunk 1 above)
Comment (adam-maj) [reply src/schema/tokens/pack.ts:16]: I think quantityPerReward is an important distinction here - added it because I originally had just `quantity` as the interface, but when I was using it, it felt like quantity was the number of tokens I was going to wrap (which it isn't), so it felt confusing. What I had in mind was wouldn't want to have a user put something like:
```
quantity: 1
totalRewards: 50
```
expecting to only lose 1 token, and then lose 50, so wanted to make it explicit.

File: test/pack.test.ts (hunk 4 above)
Comment (adam-maj) [reply test/pack.test.ts:105]: yep good idea, maybe should to something similar for MW

File: src/schema/tokens/pack.ts (hunk 2 above)
Comment (adam-maj) [reply src/schema/tokens/pack.ts:67]: Schema is a bit different for rewards vs MW because of the amounts stuff

File: src/contracts/pack.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (joaquim-verges) [reply src/contracts/pack.ts:472]: i see - that makes sense.

File: src/contracts/pack.ts [hunk 8]
```diff
@@ -121,448 +125,428 @@ export class Pack implements UpdateableNetwork {
...
-          this.contractWrapper.getProvider(),
-          rewardId.toString(),
-          this.storage,
-        ),
-    }));
+  /**
+   * Create Pack
+   * @remarks Create a new pack with the given metadata and rewards and mint it to the connected wallet.
+   *
+   * @param metadataWithRewards - the metadata and rewards to include in the pack
```
Comment (joaquim-verges) [inline src/contracts/pack.ts:275]: let's add a `See {@link Pack.createTo}` in the remarks

File: src/contracts/pack.ts (hunk 5 above)
Comment (joaquim-verges) [reply src/contracts/pack.ts:291]: fix the example after the `metadata` rename?
